src/transformStateWithClones.js

```diff
@@ -7,7 +7,43 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesArray = [];
+
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        const keys = Object.keys(currentState);
+        const keysToRemove = keys.filter((key) => action.keysToRemove.includes(key));
+
+        const newState = {};
+
+        keys.forEach((key) => {
+          if (!keysToRemove.includes(key)) {
+            newState[key] = currentState[key];
+          }
+        });
+
+        currentState = { ...newState }
```

Comment:
The logic for removing properties is more complex than necessary. You can simplify the removal process by directly deleting the keys from a copy of the current state instead of creating a new object and copying over the keys that are not to be removed.